hathor/cli/db_export.py

```diff
@@ -0,0 +1,99 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import io
+import struct
+from argparse import ArgumentParser, FileType, Namespace
+from typing import Iterator, Optional
+
+from hathor.cli.run_node import RunNode
+
+MAGIC_HEADER = b'HathDB'
+
+
+class DbExport(RunNode):
+    def start_manager(self, args: Namespace) -> None:
+        pass
+
+    def register_signal_handlers(self, args: Namespace) -> None:
+        pass
+
+    def register_resources(self, args: Namespace) -> None:
+        pass
+
+    def create_parser(self) -> ArgumentParser:
+        from hathor.conf import HathorSettings
+        settings = HathorSettings()
+
+        def max_height(arg: str) -> Optional[int]:
+            if arg.lower() == 'checkpoint':
+                if not settings.CHECKPOINTS:
+                    raise ValueError('There are no checkpoints to use')
+                return settings.CHECKPOINTS[-1].height
+            elif arg:
+                return int(arg)
+            else:
+                return None
+
+        parser = super().create_parser()
+        parser.add_argument('--export-file', type=FileType('wb', 0), required=True,
+                            help='Save the export to this file')
+        parser.add_argument('--export-iterator', choices=['fast2', 'fast', 'slow'], default='fast2',
+                            help='Which method of iterating to use, don\'t change unless you know what it does')
+        parser.add_argument('--export-max-height', type=max_height,
+                            help='Stop exporting after a certain height, use "checkpoint" stop at the last one')
+        return parser
+
+    def prepare(self, args: Namespace) -> None:
+        from hathor.transaction import BaseTransaction
+
+        super().prepare(args)
+
+        # allocating io.BufferedWriter here so we "own" it
+        self.out_file = io.BufferedWriter(args.export_file)
+
+        self.iter_tx: Iterator[BaseTransaction]
+        if args.export_iterator == 'fast2':
+            self.iter_tx = self.tx_storage._topological_fast2()
+        elif args.export_iterator == 'fast':
+            self.iter_tx = self.tx_storage._topological_fast()
+        elif args.export_iterator == 'slow':
+            self.iter_tx = self.tx_storage._topological_sort()
+        else:
+            raise ValueError(f'unknown iterator "{args.export_iterator}"')
+
+        self.export_height = args.export_max_height
+
+    def run(self) -> None:
+        from hathor.util import progress
+        self.out_file.write(MAGIC_HEADER)
+        self.log.info('export')
+        for tx in progress(self.iter_tx, log=self.log):
+            assert tx.hash is not None
+            tx_meta = tx.get_metadata()
+            # write tx
+            if tx.is_genesis:
+                continue
+            tx_bytes = bytes(tx)
+            self.out_file.write(struct.pack('!I', len(tx_bytes)) + tx_bytes)
```

Comment:
I feel we should call write twice instead of concatenating `len(tx_bytes)` with `tx_bytes` before doing it. Such a concatenation will require memory allocation and two memcpy.

Reply:
You're right, although it's a non-issue currently because iterating is several orders of magnitudes slower, so this doesn't make much of a difference.

Reply:
Thanks for doing it.